userspace/libscap/engine/bpf/scap_bpf.c

```diff
@@ -835,13 +833,29 @@ static int32_t load_bpf_file(
 	return res;
 }
 
-static void *perf_event_mmap(struct bpf_engine *handle, int fd, uint32_t *size)
+static void *perf_event_mmap(struct bpf_engine *handle, int fd, uint32_t *size, uint64_t buf_num_pages)
 {
+   /* Each data page is mapped twice to allow "virtual"
+	* continuous read of samples wrapping around the end of ring
+	* buffer area:
+	* 
+	* ------------------------------------------------------
+	* | meta pages |  real data pages  |  same data pages  |
+	* ------------------------------------------------------
+	* |            | 1 2 3 4 5 6 7 8 9 | 1 2 3 4 5 6 7 8 9 |
+	* ------------------------------------------------------
+	* |            | TA             DA | TA             DA |
+	* ------------------------------------------------------
+	*                               ^^^^^^^
+	*                                  |
+	* Here, no need to worry about special handling of wrapped-around
+	* data due to double-mapped data pages.
+	*/
 	int page_size = getpagesize();
-	int ring_size = page_size * BUF_SIZE_PAGES;
+	int ring_size = page_size * buf_num_pages;
```

Comment:
The bpf_map_create params still come from the ELF file, don't they?

Reply:
yeah but it doesn't contain the buffer dimension, so I don't see the problem :thinking:

Reply:
OK, I probably misread the code. But in that case, does the mmap determine the size of the underlying map? I'm thoroughly confused. How does e.g. max_entries (https://github.com/falcosecurity/libs/blob/master/userspace/libscap/engine/bpf/scap_bpf.c#L199) come into play here?

We're changing the mmapped size but I can't see anything about changing the actual underlying eBPF map. Can you point me to the place where it's determined?

Reply:
`max_entries` is the number of buffers we want so one for  every CPU, we patch the elf value with the number of CPUs here https://github.com/falcosecurity/libs/blob/master/userspace/libscap/engine/bpf/scap_bpf.c#L386-L392.

Yeah the flow is:
1. you create the BPF map (updating the `max_entries` to the number of CPU) https://github.com/falcosecurity/libs/blob/master/userspace/libscap/engine/bpf/scap_bpf.c#L386-L392
2. you obtain the `fd` with `sys_perf_event_open` https://github.com/falcosecurity/libs/blob/464da56d3caba7f811ad564c035ac2c5f8f972aa/userspace/libscap/engine/bpf/scap_bpf.c#L1550
3. you mmap the fd https://github.com/falcosecurity/libs/blob/464da56d3caba7f811ad564c035ac2c5f8f972aa/userspace/libscap/engine/bpf/scap_bpf.c#L860
4. you update the BPF map with the `fd` https://github.com/falcosecurity/libs/blob/464da56d3caba7f811ad564c035ac2c5f8f972aa/userspace/libscap/engine/bpf/scap_bpf.c#L1559

(2),(3),(4) are per-CPU actions

Reply:
OK, so what exactly does the ring_size represent (apart from the obvious size of the mmapped buffer)? How does changing the mmap window size affect anything? If the only thing passed to bpf_map_update_elem is the fd number, the mmap VMA size cannot be used for anything (there are multiple mappings and they could trivially have different sizes).